Update Jakarta API bundles

tevans78 opened the pull request:
Ensure that all Jakarta API bundles are using exactly the right upstream version and include all ancillary files such as licence and notices.

- Ensure that all references to upstream artifacts are done using an exact version
- Use -includeresource rather than Include-Resource
- Include all additional (non-class) files such as licenses and notices
- Remove unneeded elements from the buildpath

tevans78 committed:
Update Jakarta API bundles

tevans78 commented:
#build

tevans78 commented:
#build

tevans78 commented:
#build

tevans78 commented:
#build

jhanders34 commented on dev/io.openliberty.jakarta.expressionLanguage.4.0/bnd.bnd:
```diff
@@ -24,7 +24,8 @@ Export-Package: \
 
 Include-Resource: \
-    jakarta=resources/jakarta
+    jakarta=resources/jakarta, \
+    @${repo;org.apache.tomcat:tomcat-el-api;10.0.18;EXACT}!/!(META-INF/maven/*|module-info.class)
```
will this get rid of overlay?  Does the order need to get flipped?

tevans78 replied:
Does not appear to but probably doesn't hurt to flip it anyway.

jhanders34 commented on dev/io.openliberty.jakarta.expressionLanguage.5.0/bnd.bnd:
```diff
@@ -29,7 +29,8 @@ Export-Package: \
 
 Include-Resource: \
-    jakarta=resources/jakarta
+    jakarta=resources/jakarta, \
+    @${repo;org.apache.tomcat:tomcat-el-api;10.1.2;EXACT}!/!(META-INF/maven/*|module-info.class)
```
Same flip question about overlay.

tevans78 replied:
Same answer - Does not appear to be a problem but probably doesn't hurt to flip it anyway.

jhanders34 reviewed (changes requested):
1.  In general `-includeresource` is the new way of doing `Include-Resource`.  It would be preferred to not use the old one if you are going to be adding new to a file.  It would be good to eventually convert all of them that we can convert, but some will not be possible to convert due to apache issues.
2. Should io.openliberty.jakarta/cdi.4.0.bnd also be updated to use exact instead of range?
3. Removing the buildpath entries in some of these bnd files will cause the import version ranges to now go away.  Did you intend for that?

tevans78 commented:
> 1. In general `-includeresource` is the new way of doing `Include-Resource`.  It would be preferred to not use the old one if you are going to be adding new to a file.  It would be good to eventually convert all of them that we can convert, but some will not be possible to convert due to apache issues.

I will attempt to convert where possible

> 2. Should io.openliberty.jakarta/cdi.4.0.bnd also be updated to use exact instead of range?

Yes

> 3. Removing the buildpath entries in some of these bnd files will cause the import version ranges to now go away.  Did you intend for that?

No ... but now I'm not sure which way is "correct". If import version ranges are required then perhaps they should be explicitly specified rather than relying on the buildpath?

tevans78 commented:
#build

tevans78 commented:
@jhanders34 I think I have addressed the issues above.
1. All of the bundles changed in this PR now use `-includeresource`. I checked the manifests of the upstream artifacts and I did not find any that had a `Include-Resource` header.
2. I removed the version range from the cdi-4.0 bundle
3. In the cases where the buildpath affected the import version ranges, I added those entries back in*.

*We are generally very inconsistent about how we handle those import version ranges. Some bundles pick up the ranges via the buildpath, some hardcode the ranges and some do not specify ranges at all. As we discussed on slack at the end of last year, it would be nice to standardize this, particularly if we can then move more of these bundles into the single `io.openliberty.jakarta` project which does not use a buildpath at all. That should be a separate PR.

Although I have a good enough PB, I probably can't merge this just yet because there will be a clash with the update to myfaces 4.0.0.RC3. I will have to re-merge once that is delivered.

jhanders34 commented on dev/io.openliberty.jakarta.authentication.2.0/bnd.bnd:
```diff
@@ -23,6 +23,8 @@ instrument.disabled: true
 publish.wlp.jar.suffix: dev/api/spec
 
+-includeresource: \
+  @${repo;jakarta.authentication:jakarta.authentication-api;2.0.0;EXACT}!/!(META-INF/maven/*|module-info.class)
```
```suggestion
  @${repo;jakarta.authentication:jakarta.authentication-api;2.0.0;EXACT}!/!(META-INF/maven/*|module-info.class|exclude-common.xml)
```

tevans78 replied:
done

jhanders34 commented on dev/io.openliberty.jakarta.authorization.2.0/bnd.bnd:
```diff
@@ -20,5 +20,8 @@ instrument.disabled: true
 publish.wlp.jar.suffix: dev/api/spec
 
+-includeresource: \
+  @${repo;jakarta.authorization:jakarta.authorization-api;2.0.0;EXACT}!/!(META-INF/maven/*|module-info.class)
```
```suggestion
  @${repo;jakarta.authorization:jakarta.authorization-api;2.0.0;EXACT}!/!(META-INF/maven/*|module-info.class|exclude-common.xml)
```

tevans78 replied:
done

jhanders34 commented on dev/io.openliberty.jakarta.connectors.2.0/bnd.bnd:
```diff
@@ -21,8 +21,8 @@ Export-Package: \
-Include-Resource: \
-  @${repo;jakarta.resource:jakarta.resource-api;2.0.0;EXACT}!/META-INF/NOTICE
+-includeresource: \
+  @${repo;jakarta.resource:jakarta.resource-api;2.0.0;EXACT}!/!(META-INF/maven/*|module-info.class)
```
```suggestion
  @${repo;jakarta.resource:jakarta.resource-api;2.0.0;EXACT}!/!(META-INF/maven/*|module-info.class|doc-files/*)
```

tevans78 replied:
Are you sure about this one? The only file in `doc-files` is `speclicense.html` and part of the aim of this PR was to _include_ such license files.

jhanders34 replied:
META-INF/LICENSE is in the bundle now.  I figured that was sufficient.  I assumed the doc files was for just that doc.  I didn't think it belonged in the bundle, but I can live with it being there as well.

tevans78 replied:
Will make the suggested change.

tevans78 replied:
done

jhanders34 commented on dev/io.openliberty.jakarta.messaging.3.0/bnd.bnd:
```diff
@@ -8,9 +8,18 @@
+-include= ~../cnf/resources/bnd/bundle.props
+bVersion=1.0
+
+Bundle-SymbolicName: io.openliberty.jakarta.messaging.3.0; singleton:=true
```
```suggestion
Bundle-SymbolicName: io.openliberty.jakarta.messaging.3.0; singleton:=true
Bundle-Description: Jakarta Messaging 3.0 Specification
Bundle-Name: Jakarta Messaging API

Specification-Version: 3.0
```

The specification version is needed for knowing what version is being used for one of the JMS APIs.  An EE9 repeat fat likely fails without this update and the EE 9 CTS will fail without this change as well.  See #21464 for details.

tevans78 replied:
done

tevans78 commented:
- All review updates made.
- Rebased to pick up copyright header changes
- Updated copyright date to 2023

tevans78 commented:
#build

tevans78 commented:
#libby
